### PR TITLE
SDK-2955 Ensure issuer and audience validation are required

### PR DESCRIPTION
**Pull request by sergiughf**
## Description

In this PR we ensure that issuer and audience validation happen out of the box. 

## Testing

- [x] This change adds test coverage for new/changed/fixed functionality

## Checklist

- [x] I have read and agreed to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).
- [x] I have read the [Auth0 General Contribution Guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md).
- [x] I have reviewed my own code beforehand.
- [ ] I have added documentation for new/changed functionality in this PR.
- [ ] All active GitHub checks for tests, formatting, and security are passing.
- [ ] The correct base branch is being used, if not `master`.

### CODE REVIEW

File: validate/josev2/validator.go [hunk 1]
```diff
@@ -14,21 +14,11 @@ import (
 type Validator struct {
 	keyFunc            func(context.Context) (interface{}, error) // Required.
 	signatureAlgorithm jose.SignatureAlgorithm                    // Required.
```
Comment (Widcket) [inline validate/josev2/validator.go:16]: Can we do this without depending on `jose.SignatureAlgorithm`, so we could swap it out in the future if needed, e.g. requiring a `string` instead?

Comment (Widcket) [reply validate/josev2/validator.go:16]: Or providing our own set of valid constants?

File: validate/josev2/validator.go [hunk 2]
```diff
@@ -14,21 +14,11 @@ import (
...
-	expectedClaims     func() jwt.Expected                        // Optional.
-	customClaims       func() CustomClaims                        // Optional.
+	expectedClaims     jwt.Expected                               // Optional.
```
Comment (Widcket) [inline validate/josev2/validator.go:17]: Same as above, can we do this without requiring explicitly the `jwt.Expected` type?

File: validate/josev2/validator_test.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+			name:  "it throws an error when it fails to fetch the keys from the key func",
+			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL2dvLWp3dC1taWRkbGV3YXJlLmV1LmF1dGgwLmNvbS8iLCJzdWIiOiIxMjM0NTY3ODkwIiwiYXVkIjpbImh0dHBzOi8vZ28tand0LW1pZGRsZXdhcmUtYXBpLyJdfQ.-R2K2tZHDrgsEh9JNWcyk4aljtR6gZK0s2anNGlfwz0",
+			keyFunc: func(context.Context) (interface{}, error) {
+				return nil, errors.New("key func error message")
+			},
+			expectedError: errors.New("error getting the keys from the key func: key func error message"),
+		},
+		{
+			name:  "it throws an error when it fails to deserialize the claims",
+			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL2dvLWp3dC1taWRkbGV3YXJlLmV1LmF1dGgwLmNvbS8iLCJzdWIiOiIxMjM0NTY3ODkwIiwiYXVkIjpbImh0dHBzOi8vZ28tand0LW1pZGRsZXdhcmUtYXBpLyJdfQ.vR2K2tZHDrgsEh9zNWcyk4aljtR6gZK0s2anNGlfwz0",
```
Comment (Widcket) [inline validate/josev2/validator_test.go:97]: Just a question, why in this particular case the claims fail to be deserialized, when they seem correct (syntax & expected type)?
The error seems to be related to the signature.

File: validate/josev2/validator.go [hunk 4]
```diff
@@ -42,21 +32,30 @@ type UserContext struct {
...
 	signatureAlgorithm jose.SignatureAlgorithm,
+	issuerURL string,
+	audience jwt.Audience,
 	opts ...Option,
 ) (*Validator, error) {
 	if keyFunc == nil {
 		return nil, errors.New("keyFunc is required but was nil")
 	}
+	if signatureAlgorithm == "" {
+		return nil, errors.New("signature algorithm is required but was empty")
```
Comment (Widcket) [inline validate/josev2/validator.go:43]: Are these errors tested?

**Review by Widcket — COMMENTED** [review]
Looking good, just have a couple of questions.

File: validate/josev2/validator_test.go (hunk 3 above)
Comment (sergiughf) [reply validate/josev2/validator_test.go:96]: I updated the test name to better reflect why we fail to deserialize the claims.

Comment (sergiughf) [reply validate/josev2/validator_test.go:96]: https://github.com/auth0/go-jwt-middleware/pull/119/commits/14e3dc69a6f743b40504778ca4788446aac038dd

File: validate/josev2/validator.go (hunk 4 above)
Comment (sergiughf) [reply validate/josev2/validator.go:43]: Good catch, I added them here https://github.com/auth0/go-jwt-middleware/pull/119/commits/b7f98832bad0a6ed7d76a1efd6d9095d907c86dc

File: validate/josev2/validator_test.go (hunk 3 above)
Comment (sergiughf) [reply validate/josev2/validator_test.go:96]: Also changed the signature of the validator.New to not use any jose pkg structs.

**Review by Widcket — APPROVED** [review]
LGTM

**Pull request merged**